aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -375,4 +367,137 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
   return result;
 }
 
+template <typename scalar_t, bool inplace, bool upper>
+static void apply_triu_tril_single(
+    scalar_t* result, scalar_t* self,
+    int64_t k, int64_t n, int64_t m,
+    int64_t res_row_stride, int64_t res_col_stride,
+    int64_t self_row_stride, int64_t self_col_stride) {
+
+  constexpr int64_t zero = 0;
+  int64_t i;
+
+  if (upper) {
+    #pragma omp parallel for private(i)
+    for (i = 0; i < n; i++) {
+      for (int64_t j = 0; j < std::min(m, i + k); j++) {
+        result[i * res_row_stride + j * res_col_stride] = 0;
+      }
+      if (!inplace) {  // copy the rest of the self if not inplace
+        for (int64_t j = std::max(zero, i + k); j < m; j++) {
+          result[i * res_row_stride + j * res_col_stride] = self[i * self_row_stride + j * self_col_stride];
+        }
+      }
+    }
+  } else {
+    #pragma omp parallel for private(i)
+    for (i = 0; i < n; i++) {
+      for (int64_t j = std::max(zero, i + k + 1); j < m; j++) {
+        result[i * res_row_stride + j * res_col_stride] = 0;
+      }
+      if (!inplace) {  // copy the rest of the self if not inplace
+        for (int64_t j = zero; j < std::min(m, i + k + 1); j++) {
+          result[i * res_row_stride + j * res_col_stride] = self[i * self_row_stride + j * self_col_stride];
+        }
+      }
+    }
+  }
+}
+
+template <typename scalar_t, bool inplace, bool upper>
+void apply_triu_tril(Tensor& result, const Tensor& self, int64_t k) {
+  auto n = self.size(-2);
+  auto m = self.size(-1);
+  auto self_data = self.data<scalar_t>();
+  auto self_stride = self.dim() > 2 ? self.stride(-3) : n * m;
+  auto batchsize = batchCount(self);
+  auto self_row_stride = self.stride(-2);
+  auto self_column_stride = self.stride(-1);
+
+  auto result_data = result.data<scalar_t>();
+  int64_t result_stride, result_row_stride, result_column_stride;
+  if (result_data != self_data) {
+    result_stride = result.dim() > 2 ? result.stride(-3) : n * m;
+    result_row_stride = result.stride(-2);
+    result_column_stride = result.stride(-1);
+  } else {
+    result_stride = self_stride;
+    result_row_stride = self_row_stride;
+    result_column_stride = self_column_stride;
+  }
+
+  int64_t b;
+
+  #pragma omp parallel for private(b)
+  for (b = 0; b < batchsize; b++) {
```

Comment:
Yes, you are correct, thanks for catching that. I don't think there's an easy way to iterate over the batches. One way we could resolve the problem is call `tensor.contiguous()` if the batch dimensions are non-contiguous and then the rest of the code should work.

The alternative approach would be to convert a linear index (that ranges from 0 to `tensor.numel()`) to an index vector, then do a dot product with the strides to get the memory location.

Reply:
I think the former approach would be relatively cheaper, given that the dot product would have to be computed multiple times.

While analyzing this problem, I also thought of a mask based solution. While I am not sure if it would be faster, I am certain that it would retain the (non-)contiguity of the input tensor. We can create a mask using `apply_triu_tril_single`, and then expand it to the required size, after which `at::where` can be used to obtain the result.

Reply:
I agree that the first approach is cheaper and would probably be less overhead for non-batched triu/tril. Does the same problem exist on the CUDA side? It might be a little strange if the CUDA operator doesn't make the tensor contiguous in some cases when the CPU operator does

Reply:
This is not properly handled on the cuda side at all https://github.com/pytorch/pytorch/pull/15257/files#diff-5b3ae8ff75409dce3f752955b36b9359R433
1) For a batched case, all batch dimensions are assumed to be contiguous (batch_stride = tensor.stride(-3))
2) For a non-batched case batch stride is calculated as rows*cols, which is not right, but probably benign.

Reply:
I am sorry, I will fix this case as soon as I can, and keep you posted.

As suggested, I will make the tensor contiguous only if the batch dims are non-contiguous.